### DISCUSSION TITLE
Improvement: Add ObjectLifeCycle pattern methods to all structs

### DISCUSSION THREAD

**Discussion by OmidHekayati** [pactus-project/pactus#544] [Ideas]
- When All structs have ObjectLifeCycle methods(`Init()`, `Reinit()`, `Deinit()`), All developers can easily track the object life cycle.
- If an object needs huge allocation and prefers to reuse after the first usage (with proper benchmarks), we can easily reuse them
- If a struct needs to be embedded in another struct, we always track easily how and why! e.g. I can't understand why [state.LoadOrNewState()](https://github.com/pactus-project/pactus/blob/a75fe24372bd3cd35d6a6c0d725f26e84a1c0357/state/state.go#L48) return `Facade` interface type instead of a pointer to its type as `state` and even why `state` type is internal to its package! It is tough to track all usage of returned `Facade` and how others use or keep a reference to it and also hard to make suggestions to improve performance.

**Comment by OmidHekayati** [comment]
In other words, API must be flexible. Accepting Interfaces and Returning Structs.

[Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) say:
`be conservative in what you do, and be liberal in what you accept from others.`

**Comment by b00f** [comment]
In large modules like `state`, we have defined an interface (like `Facade` in state). 
By following this approach, we can easily mock a large object (like `state`) for testing purposes.

**Reply by OmidHekayati** [reply]
`state` can implement `Facade`, I ask why [`state.LoadOrNewState()`](https://github.com/pactus-project/pactus/blob/a75fe24372bd3cd35d6a6c0d725f26e84a1c0357/state/state.go#L48) return `Facade` interface? you can use the `state` pointer anywhere that needs `Facade`!
If it is a test that always sure `state` implement the `Facade`, You can easily add the below simple one-line code to `state_test.go`
```go
var _ Facade = &state{}
```

Also, we can use Go build tags `go:build` to mock inside a package for test purposes to prevent runtime performance impact. So `package state` can provide a fake state object with `go:build test` flag! We have always a way to improve performance without reducing code readability.

**Reply by b00f** [reply]
@OmidHekayati I'm not sure if I've understood correctly. You suggest to change:

`func LoadOrNewState(...) (Facade, error)`
to
`func LoadOrNewState(...) (*state, error)`

@kehiy could you please check if we can return `*state`, even I have doubt, because it is unreported. 

Regarding mocks, I know there are other ways to mock the objects, but here we have better control over what exactly the mock should respond. However, we can refactor as you mentioned, if it gives us better and simpler tests.

**Reply by OmidHekayati** [reply]
Yes, You get it rightly. When using this signature format like `func LoadOrNewState(...) (*state, error)` can easily assign to any `Facade` variable type, So it is not breaking any codes. It can be the start to improve performance to emit unnecessary interface usage.

**Reply by b00f** [reply]
@kehiy go ahead man! Let's make @OmidHekayati happy!

**Reply by kehiy** [reply]
I like to make Omid happy too, but I think resolving this conversation is more than I can (maybe) :smile: 
how much time do we have?

**Reply by OmidHekayati** [reply]
@kehiy Don't bother yourself. I just suggest, So It is a COULD, not a MUST. You have infinite time to solve performance issues!

**Reply by kehiy** [reply]
yeah, that's true, for now, I have another task after it I go thorough context issues then this one...
